examples/erc1155/lib.rs

```diff
@@ -0,0 +1,196 @@
+// Copyright 2018-2021 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use ink_lang as ink;
+use ink_prelude::vec::Vec;
+
+type TokenId = u128;
+type Balance = <ink_env::DefaultEnvironment as ink_env::Environment>::Balance;
+
+#[ink::trait_definition]
+pub trait Erc1155 {
+    #[ink(message)]
+    fn safe_transfer_from(&mut self);
+
+    #[ink(message)]
+    fn safe_batch_transfer_from(&mut self);
+
+    #[ink(message)]
+    fn balance_of(&self, owner: ink_env::AccountId, token_id: TokenId) -> Balance;
+
+    #[ink(message)]
+    fn balance_of_batch(
+        &self,
+        owners: Vec<ink_env::AccountId>,
+        token_ids: Vec<TokenId>,
+    ) -> Vec<Balance>;
+
+    #[ink(message)]
+    fn set_approval_for_all(&mut self);
+
+    #[ink(message)]
+    fn is_approved_for_all(&self);
+}
+
+#[ink::trait_definition]
+pub trait Erc1155TokenReceiver {
+    #[ink(message)]
+    fn on_erc_1155_received(&mut self);
+
+    #[ink(message)]
+    fn on_erc_1155_batch_received(&mut self);
+}
+
+#[ink::contract]
+mod erc1155 {
+    use super::*;
+
+    use ink_prelude::collections::BTreeMap;
+
+    /// An ERC-1155 contract.
+    #[ink(storage)]
+    pub struct Contract {
+        balances: BTreeMap<AccountId, BTreeMap<TokenId, Balance>>,
```

Comment:
We usually prefer to use something like `BTreeMap<(AccountId, TokenID), Balance>` instead.